### ISSUE TITLE
What does "default copy constructible" mean?

### ISSUE THREAD

**Issue by psalz** [KhronosGroup/SYCL-Docs#210]
Section **4.3.3 Common by-value semantics** currently states that types such as `id` and `range` must be "*default* copy constructible", "*default* destructible", "*default* move constructible" and "*default* move assignable". What does that mean? As far as I can tell, these terms are not used anywhere on cppreference.

I assume the intent was to say *trivially* constructible/destructible/moveable. This would already be true for most implementations and types (hipSYCL does not fulfill all requirements for `item`, `h_item` and `nd_range`).

cc @AerialMantis I think you came up with this wording originally in 2e3b42c7.

**Comment by AerialMantis** [comment]
Yes I think you're right, so the original intent of this wording was to require the types to use the implicit default constructor, move/copy constructors and destructor, so as I understand this would translate to `is_trivially_default_constructible`, `is_trivially_copy_constructible`, `is_trivially_move_constructible` and `is_trivially_destructible`, I think this would better describe the intent than the current wording.

**Comment by psalz** [comment]
As I understand it `is_trivially_*` is actually a stronger guarantee. For example a type `struct foo { std::vector<int> bar; };` has an implicitly defaulted copy constructor, but is not trivially copy-constructible.

After looking some more into this I'm actually a bit confused:

Section 4.5.3 says special member functions should be defaulted. However, it also lists all of them in table 9 and says that classes must either follow the rule of zero or the rule of five. The latter implies that there can also be explicitly defined special member functions, which means they are no longer defaulted and classes are no longer trivially-copyable.

I'm pretty sure at least `sycl::id` and `sycl::range` are supposed to be device-copyable, right? Assuming they *may* follow the rule of five and are therefore not guaranteed to be trivially-copyable, they would have to be listed explicitly as device-copyable in section 3.13.1, which they are not.

**Comment by gmlueck** [comment]
> I'm pretty sure at least sycl::id and sycl::range are supposed to be device-copyable, right?

I propose that we define these as trivially copyable in https://gitlab.khronos.org/sycl/Specification/-/issues/564

**Comment by fraggamuffin** [comment]
Linked to internal github issue on marray improvements

**Comment by fraggamuffin** [comment]
leaning towards changing it to trivially_#
